src/Servers/HttpSys/src/RequestProcessing/Response.cs

```diff
@@ -289,45 +302,57 @@ private void EnsureResponseStream()
 
         uint statusCode;
         uint bytesSent;
-        List<GCHandle>? pinnedHeaders = SerializeHeaders(isOpaqueUpgrade);
-        try
+        using UnmanagedBufferWriter writer = new(4096);
+        SerializeHeaders(isOpaqueUpgrade, writer);
+
+        if (dataChunks != null)
         {
-            if (dataChunks != null)
-            {
-                if (pinnedHeaders == null)
-                {
-                    pinnedHeaders = new List<GCHandle>();
-                }
-                var handle = GCHandle.Alloc(dataChunks, GCHandleType.Pinned);
-                pinnedHeaders.Add(handle);
-                _nativeResponse.Response_V1.EntityChunkCount = (ushort)dataChunks.Length;
-                _nativeResponse.Response_V1.pEntityChunks = (HttpApiTypes.HTTP_DATA_CHUNK*)handle.AddrOfPinnedObject();
-            }
-            else if (asyncResult != null && asyncResult.DataChunks != null)
-            {
-                _nativeResponse.Response_V1.EntityChunkCount = asyncResult.DataChunkCount;
-                _nativeResponse.Response_V1.pEntityChunks = asyncResult.DataChunks;
-            }
-            else
-            {
-                _nativeResponse.Response_V1.EntityChunkCount = 0;
-                _nativeResponse.Response_V1.pEntityChunks = null;
-            }
+            _nativeResponse.Response_V1.EntityChunkCount = (ushort)dataChunksCount;
+            _nativeResponse.Response_V1.pEntityChunks = dataChunks;
+        }
+        else if (asyncResult != null && asyncResult.DataChunks != null)
+        {
+            _nativeResponse.Response_V1.EntityChunkCount = asyncResult.DataChunkCount;
+            _nativeResponse.Response_V1.pEntityChunks = asyncResult.DataChunks;
+        }
+        else
+        {
+            _nativeResponse.Response_V1.EntityChunkCount = 0;
+            _nativeResponse.Response_V1.pEntityChunks = null;
+        }
 
-            var cachePolicy = new HttpApiTypes.HTTP_CACHE_POLICY();
-            if (_cacheTtl.HasValue && _cacheTtl.Value > TimeSpan.Zero)
-            {
-                cachePolicy.Policy = HttpApiTypes.HTTP_CACHE_POLICY_TYPE.HttpCachePolicyTimeToLive;
-                cachePolicy.SecondsToLive = (uint)Math.Min(_cacheTtl.Value.Ticks / TimeSpan.TicksPerSecond, Int32.MaxValue);
-            }
+        var cachePolicy = new HttpApiTypes.HTTP_CACHE_POLICY();
+        if (_cacheTtl.HasValue && _cacheTtl.Value > TimeSpan.Zero)
+        {
+            cachePolicy.Policy = HttpApiTypes.HTTP_CACHE_POLICY_TYPE.HttpCachePolicyTimeToLive;
+            cachePolicy.SecondsToLive = (uint)Math.Min(_cacheTtl.Value.Ticks / TimeSpan.TicksPerSecond, Int32.MaxValue);
+        }
 
-            byte[] reasonPhraseBytes = HeaderEncoding.GetBytes(reasonPhrase);
-            fixed (byte* pReasonPhrase = reasonPhraseBytes)
+        Span<byte> reasonPhraseBytes = HeaderEncoding.GetBytes(reasonPhrase, writer);
+        fixed (byte* pReasonPhrase = reasonPhraseBytes)
```

Comment:
This does not need to use `fixed`. It can just be `AsPointer` like in the other places.

It may be better if GetBytes signature is `byte* GetBytes(string s, out int length, UnmanagedBufferWriter writer)`  because all callers want the pointer at the end. It would avoid duplication of the `Unsafe.AsPointer(ref Memory.Marshal(...` dance everywhere.

Reply:
GetBytes can be even instance method on UnmanagedBufferWriter.